nouveau/src/main/java/org/apache/couchdb/nouveau/core/IndexManager.java

```diff
@@ -82,33 +114,97 @@ public interface IndexFunction<V, R> {
+            holder.lock.readLock().lock();
 
-            final CompletableFuture<Index> future = cache.get(name);
-            final Index index = future.join();
-
-            if (index.tryAcquire()) {
+            // Load if not already loaded.
+            if (holder.state == HolderState.NOT_LOADED) {
+                holder.lock.readLock().unlock();
+                holder.lock.writeLock().lock();
                 try {
-                    return indexFun.apply(index);
+                    if (holder.state == HolderState.NOT_LOADED) {
+                        holder.index = load(name);
+                        holder.commitFuture = this.schedulerExecutorService.scheduleWithFixedDelay(
+                                commitFun(name, holder),
+                                commitIntervalSeconds,
+                                commitIntervalSeconds,
+                                TimeUnit.SECONDS);
+                        holder.state = HolderState.LOADED;
+                    }
+                    holder.lock.readLock().lock();
                 } finally {
-                    index.release();
+                    holder.lock.writeLock().unlock();
                 }
             }
+
+            try {
+                switch (holder.state) {
+                    case NOT_LOADED:
+                        throw new IllegalStateException();
+                    case UNLOADED:
+                        Thread.sleep(1000);
+                        continue retry;
+                    case LOADED:
+                        return indexFun.apply(holder.index);
+                }
+            } finally {
+                holder.lock.readLock().unlock();
+            }
+        }
+    }
+
+    public void unload(final String name, final boolean forceDelete) throws IOException, InterruptedException {
+        final IndexHolder holder;
+        synchronized (cache) {
+            holder = cache.computeIfAbsent(name, (k) -> new IndexHolder());
+        }
+        holder.lock.writeLock().lock();
+        try {
+            switch (holder.state) {
+                case LOADED:
+                    if (forceDelete) {
+                        holder.index.setDeleteOnClose(true);
+                    }
+                    LOGGER.info("closing {}", name);
+                    try {
+                        close(name, holder);
+                    } catch (final IOException e) {
+                        LOGGER.error("I/O exception when evicting {}", name, e);
+                    }
+                    holder.state = HolderState.UNLOADED;
+                    holder.index = null;
+                    synchronized (cache) {
+                        cache.remove(name, holder);
+                    }
+                    break;
+                case NOT_LOADED:
+                case UNLOADED:
+                    if (forceDelete) {
+                        IOUtils.rm(indexRootPath(name));
+                    }
```

Comment:
Are there any assertions we can make about the presence in cache in these NOT_LOADED/UNLOADED state? For instance, would it make sense to assert we're not in the cache at this point?

Reply:
hm, perhaps. I am also pondering if we need three states. It is important not to block reading the cache while we load or unload (as these can take many seconds), hence I let you retrieve from the cache under just the cache lock, though it might return a Holder in NOT_LOADED or UNLOADED states. the first thread to acquire the write lock gets the chore of opening the index. Perhaps I really don't need a distinct UNLOADED state? It could go back to UN_LOADED, and then we'd try to load it again, but only if someone has acquired the Holder before it is removed from the cache map itself...  Hm, but no, because then we'd have to reinsert into the map. I think a boring lifecycle is better, NOT_LOADED -> LOADED -> UN_LOADED (and removed from the cache), and repeat. What do you think?